File: scala/BinaryTree.scala
import scala.collection.mutable.ArrayBuffer

object BinaryTree {
   // DFS traversal
   // - inorder = left-Root-Right
   // - preorder = Root-Left-Right
   // - postorder = Left-Right-Root 

   // DFS vs BFS
   // time complexity is the same O(n)
   // DFS requires extra space due to function call stack (recursive overhead) - O(h) where h is the depth of the graph   
   // If in our task the solution is most likely close to the root - use BFS - if close to leaf - DFS

  case class Tree(value:Integer, left:Option[Tree], right:Option[Tree])

  // time complexity is O(n) where n is number of nodes
  def breadthFirstLevels(head: Tree): Array[Array[Int]] = {
    var queue = scala.collection.mutable.Queue[Tree](head)
    val res = ArrayBuffer[Array[Int]]()
    while (!queue.isEmpty) {
        var level = Array[Int]()
        val nextLevel = scala.collection.mutable.Queue[Tree]()
        while (!queue.isEmpty) {
            val curr = queue.dequeue()
            level = level :+ curr.value
            if (curr.left.isDefined) {
                nextLevel.enqueue(curr.left.get)
            }
            if (curr.right.isDefined) {
                nextLevel.enqueue(curr.right.get)
            }
        }
        res += level
        queue = nextLevel
    } 
    res.toArray
  }

  def buildBinaryTree1ToN(n: Int): Tree = {
      var items = ArrayBuffer[Int]()
      for (i<-1 to n) {
        items += i
      }
      buildLevel(items.toArray,0).get
  }

  def printInOrder(node: Option[Tree]):Unit = {
    if (node.isDefined) {
        val currNode = node.get
        printInOrder(currNode.left)
        print(s"${currNode.value} ")
        printInOrder(currNode.right)
    }
  }

  def buildLevel(arr: Array[Int],  i:Int):Option[Tree] = {
        var root:Option[Tree] = None
        
        // Base case for recursion
        if (i < arr.length) {
            val left = buildLevel(arr, 2 * i + 1);
            val right = buildLevel(arr, 2 * i + 2);
            
            root = Some(Tree(arr(i), left, right))
        }
        root
    }

  //scalac BinaryTree.scala && scala BinaryTree
  def main(args: Array[String])={
        //val tree = Tree(1, Some(Tree(2, Some(Tree(4, None, None)), Some(Tree(5, None, None)))),Some(Tree(3, Some(Tree(6, None, None)), None)))

        val tree = buildBinaryTree1ToN(6)
        println(s"binary tree (in order) ${printInOrder(Some(tree))}")

        println(s"binary tree levels are ${breadthFirstLevels(tree).map(_.mkString(","))mkString("|")} [1|2,3|4,5,6]")
    }
}

File: java/BinaryTree.java
import java.util.*;
import java.util.stream.*;

class BinaryTree {
    // DFS traversal
    // - inorder = left-Root-Right
    // - preorder = Root-Left-Right
    // - postorder = Left-Right-Root 

    // DFS vs BFS
    // time complexity is the same O(n)
    // DFS requires extra space due to function call stack (recursive overhead) - O(h) where h is the depth of the graph   
    // If in our task the solution is most likely close to the root - use BFS - if close to leaf - DFS

    // records are available in Java 15+
    record Tree(int value, Tree left, Tree right) {}; 

    // time complexity is O(n) where n is number of nodes
    private static ArrayList<ArrayList<Integer>> breadthFirstLevels(Tree head) {
        Queue<Tree> queue = new ArrayDeque<>();
        queue.add(head);
        ArrayList<ArrayList<Integer>> res =  new ArrayList<>();
        while (!queue.isEmpty()) {
            ArrayList<Integer> level = new ArrayList<>();
            Queue<Tree> nextLevel = new ArrayDeque<>();
            while (!queue.isEmpty()) {
                Tree curr = queue.poll();
                level.add(curr.value);
                if (curr.left!=null) {
                    nextLevel.add(curr.left);
                }
                if (curr.right!=null) {
                    nextLevel.add(curr.right);
                }
            }
            res.add(level);
            queue = nextLevel;
        }
        return res;
    }

    private static Tree buildBinaryTree1ToN(int n) {
        ArrayList<Integer> items = new ArrayList<>();
        for (int i=1;i<=n; i++) {
            items.add(i);
        }

        return buildLevel(items, 0);
    }

    private static void printInOrder(Tree node) {
        if (node!=null) {
            printInOrder(node.left);
            System.out.print(node.value+ " ");
            printInOrder(node.right);
        }
    }

    private static Tree buildLevel(ArrayList<Integer> arr, int i) {
        Tree root = null;
        
        // Base case for recursion
        if (i < arr.size()) {
            Tree left = buildLevel(arr, 2 * i + 1);
            Tree right = buildLevel(arr, 2 * i + 2);
            
            root = new Tree(arr.get(i).intValue(), left, right);
        }
        return root;
    }

    // javac BinaryTree.java && java BinaryTree 
    public static void main(String[] args) {
        Tree tree = buildBinaryTree1ToN(6);
        System.out.printf("binary tree (in order) ");
        printInOrder(tree);
        System.out.println();

        ArrayList<ArrayList<Integer>> levels = breadthFirstLevels(tree);
        String res = Stream.of(levels).map(String::valueOf).collect(Collectors.joining(","));
        System.out.printf("binary tree levels are %s [[1], [2,3], [4,5,6]]\n", res);
    }
}
